src/body.cpp

```diff
@@ -902,6 +903,145 @@ void EarthAtm::SetAtmosphereHeight(double height){
+    coszens[i] = atm_coszen[i];
+  }
+  
+  unique_sort(atm_heights, height_min, height_max, height_number);
+  heights.resize(0,atm_heights.size());
+  for (unsigned int i=0; i < height_number; i++){
+    heights[i] = atm_heights[i];
+  }
+  
+  unique_sort(atm_energy, energy_min, energy_max, energy_number);
+  energies.resize(0,atm_energy.size());
+  for (unsigned int i=0; i < energy_number; i++){
+    energies[i] = atm_energy[i];
+  }
+  
+  marray<double, 3> nuEmatrix{czen_number, height_number, energy_number};
+  marray<double, 3> nuMumatrix{czen_number, height_number, energy_number};
+  
+
+  std::ifstream InputFile(filepath);
+  double cz_;
+  double h_;
+  double E_;
+  double nuEflux;
+  double nuMuflux;
+
+  while (InputFile >> cz_ >> h_ >> E_ >> nuEflux >> nuMuflux) {
+    auto it1 = std::find(coszens.begin(), coszens.end(), cz_);
+    auto it2 = std::find(heights.begin(), heights.end(), h_);
+    auto it3 = std::find(energies.begin(), energies.end(), E_);
+    if (it1 != coszens.end() && it2 != heights.end() && it3 != energies.end()) {
+      size_t czIndex = std::distance(coszens.begin(), it1);
+      size_t hIndex = std::distance(heights.begin(), it2);
+      size_t EIndex = std::distance(energies.begin(), it3);
+      nuEmatrix[czIndex][hIndex][EIndex] = nuEflux;
+      nuMumatrix[czIndex][hIndex][EIndex] = nuMuflux;
+      }
+    }
+    InputFile.close();
+    
+  Interpolators.push_back( nusquids::TriCubicInterpolator(nuEmatrix, energies, heights, coszens) );
+  Interpolators.push_back( nusquids::TriCubicInterpolator(nuMumatrix, energies, heights, coszens) );
+        
+}
+
+
+// Overrides the neutrino flux injection
+void EmittingEarthAtm::injected_neutrino_flux(marray<double, 3>& flux, const GenericTrack& track, const nuSQUIDS& nusquids) {
+      
+      if(Get_EnergyInit() == false){
+        throw std::runtime_error("Energy Space not Passed to EmittingEarthAtm instance");
+      }
+      // height calculation
+      const EarthAtm::Track& track_earthatm = static_cast<const EarthAtm::Track&>(track);
+      double czNode = track_earthatm.cosphi;
+      double xkm = track.GetX() / param.km;
+      double sinsqphi = 1 - SQR(czNode);
+      double dL = sqrt(SQR(earth_with_atm_radius) - SQR(radius) * sinsqphi) + radius * czNode;
+      double L = sqrt(SQR(radius + atm_height) - SQR(radius) * sinsqphi) - radius * czNode;
+      double r2 = SQR(earth_with_atm_radius) + SQR(xkm) - (L / param.km + dL) * xkm;
+      double r = (r2 > 0 ? sqrt(r2) : 0);
+      double rel_r = r / earth_with_atm_radius;
+      double Height = earth_with_atm_radius * (rel_r - radius / earth_with_atm_radius); // here atm_height is a member of EarthAtm in km
+      double H = Height*100000;
+
+
+      for (unsigned int ei=0; ei < ESpace.extent(0); ei++) {
+        double E = ESpace[ei]/(param.GeV);          
+        if(czNode >= czen_min && czNode<=czen_max && H>=height_min && H<=height_max && E>=energy_min && E<=energy_max){
```

Comment:
Checks on the zenith angle and height should probably be outside of the loop over energy, to reduce comparisons in the hot loop, and a single, simpler loop could be used in the case that either is out of range to zero out the entire `flux` array.